File: c/src/Structural/Decorator/ChaiDecorator.c
0


//ChaiDecorator.java - the decorator

import java.util.ArrayList;
import java.util.ListIterator;

public class ChaiDecorator extends Tea {
    private Tea teaToMakeChai;
    private ArrayList chaiIngredients = new ArrayList();
    
    public ChaiDecorator(Tea teaToMakeChai) {
        this.addTea(teaToMakeChai);
        chaiIngredients.add("bay leaf");
        chaiIngredients.add("cinnamon stick");
        chaiIngredients.add("ginger");
        chaiIngredients.add("honey");
        chaiIngredients.add("soy milk");
        chaiIngredients.add("vanilla bean");
    }

    private void addTea(Tea teaToMakeChaiIn) {
        this.teaToMakeChai = teaToMakeChaiIn;
    }
    
    public void steepTea() {
        this.steepChai();
    }

    public void steepChai() {
        teaToMakeChai.steepTea();
        this.steepChaiIngredients();
        System.out.println("tea is steeping with chai");
    }    
    
    public void steepChaiIngredients() {
        ListIterator listIterator = chaiIngredients.listIterator();
        while (listIterator.hasNext()) {
            System.out.println(((String)(listIterator.next())) + 
                                         " is steeping");
        }
        System.out.println("chai ingredients are steeping");
    }      
}

File: php/src/Structural/Decorator/ChaiDecorator.php
<?

#//ChaiDecorator.java - the decorator

require_once 'Tea.php';

class ChaiDecorator implements Tea
{

	public $teaToMakeChai;
	public $ingredients;

	function __construct($teaToMakeChai)
	{
		$this->teaToMakeChai = $teaToMakeChai;
		$this->ingredients = [
				"bay leaf",
				"cinnamon stick",
				"ginger",
				"honey",
				"soy milk",
				"vanilla bean",
		];
	}

	function steepTea ()
	{
		$this->steepChai();
	}

	function steepChai()
	{
		$this->teaToMakeChai->steepTea();
		$this->steepChaiIngredients();
	}    
   
	function steepChaiIngredients()
	{
		return implode('',$this->ingredients);
	}

}
